Unified changes done for FRR

vidyac86 opened the pull request:
Why I did it
The docker_routing_config_mode value which is set to default of separated is not working properly with the latest FRR.

How I did it
The changes are done in the docker_init.sh to modify the default behavior of considering the docker_routing_config_mode value as 'separated' when it is not seen in the config_db to modifying the value to 'unified'.

How to verify it
These values are significant to restore the values in FRR. When the value is separated, each process in FRR will have its own config file from where the configs are played back after save and restore, with the value of 'unified' there will be only one file frr.conf from where all the configs will be played back.

After doing this change, I manually verified the presence of frr.conf after save and restore and all the configs are restored properly.

Tested branch (Please provide the tested image version)
master

Tested image version
20250407.145357

Description for the changelog
Modified the default value of docker_routing_config_mode to unified.

[FRR_Unified_UT Logs.txt](https://github.com/user-attachments/files/19873691/FRR_Unified_UT.Logs.txt)

Link to config_db schema for YANG module changes
https://github.com/sonic-net/sonic-buildimage/blob/master/src/sonic-yang-models/doc/Configuration.md#device-metadata

Related PRs
https://github.com/sonic-net/sonic-utilities/pull/3840
https://github.com/sonic-net/sonic-mgmt/pulls/vidyac86

mssonicbld commented:
/azp run Azure.sonic-buildimage

venkatmahalingam commented:
The build is failing due to https://github.com/sonic-net/sonic-buildimage/pull/22538

dgsudharsan reviewed (dismissed):
we need to have full sonic-mgmt coverage to confirm these changes. Previously sonic had to move away from unified because of some issues found. We have to ensure that moving back to unified introduces no degradation.

mssonicbld commented:
/azp run Azure.sonic-buildimage

StormLiangMS commented:
We do see regression with unified mode, which lost partial config during startup in a low possibilities. The PR testing may not enough to verify this change. We should have a solid testing before get this merged.

liushilongbuaa commented:
/azpw ms_conflict

venkatmahalingam commented:
> We do see regression with unified mode, which lost partial config during startup in a low possibilities. The PR testing may not enough to verify this change. We should have a solid testing before get this merged.

As I requested earlier, Dell doesnt have an infra. to run the regression for bgpcfgd changes, need help from Microsoft to run the regression with this PR changes.

StormLiangMS commented:
/azp run

vidyac86 commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

vidyac86 commented:
/azp run Azure.sonic-buildimage

vidyac86 commented:
/azpw run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

qiluo-msft reviewed (dismissed):
as comments

vidyac86 commented on dockers/docker-fpm-frr/docker_init.sh:
```diff
@@ -56,41 +56,30 @@ if [[ ! -z "$NAMESPACE_ID" ]]; then
    update_default_gw 6
 fi
 
-if [ -z "$CONFIG_TYPE" ] || [ "$CONFIG_TYPE" == "separated" ]; then
```
Added prints to say the mode is not supported

vidyac86 committed:
Migrated the routing config mode from seperated to unified

Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

yxieca commented:
/azp run

prsunny commented:
@lguohan for viz

qiluo-msft commented on dockers/docker-fpm-frr/docker_init.sh:
```diff
@@ -56,41 +56,32 @@ if [[ ! -z "$NAMESPACE_ID" ]]; then
-        -t /usr/share/sonic/templates/gen_frr.conf.j2,/etc/frr/frr.conf \
-    "
+elif [ "$CONFIG_TYPE" == "separated" ]; then
+    echo "Config Type separated is not supported"
```
>separated is not supported [](http://example.com/codeflow?start=22&length=26)

This breaking change will break warm-reboot, because the old image config_db.json will be carried over to new image. Please update config in src/sonic-utilities/scripts/db_migrator.py, to use a supported "docker_routing_config_mode"

vidyac86 replied:
To solve this, db_migrator changes are added 
https://github.com/sonic-net/sonic-utilities/pull/3840

vidyac86 commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/AzurePipelines run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

vidyac86 committed:
Changes done to make the sonic-mgmt testsuite to run

Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

mssonicbld commented:
/azp run Azure.sonic-buildimage

venkatmahalingam commented:
@deepak-singhal0408  Are we good to merge this PR?

deepak-singhal0408 commented:
> @deepak-singhal0408 Are we good to merge this PR?

@venkatmahalingam thanks for the reminder.
Quick update: We ran the longivity tests with this PR, and the results seems to be ok.. There was a memory jump initially which stayed throughout the run.. We have started another test without this change for comparison sake.. will have results available in next 2-3 days..

With that, if all looks good, we could hopefully get this PR merged by this Friday.

vidyac86 committed:
Address FRR unified mode review comments

- Fix dead-code in docker_init.sh: remove redundant 'separated' condition
  from last elif branch as it's already handled by previous branch
- Add syslog warning for separated mode deprecation to improve visibility
  in 'show logging' output
- Rename frr_mgmt_framework_cfg_template_restore to
  use_template_render_for_restore for better clarity:
  - When true (default): use template rendering for FRR config generation
  - When false: restore from cached DB tables
- Update YANG model, sample configs, and test files to use new variable name
- Document frr.conf.j2 content equivalence in gen_frr.conf.j2 template

Signed-off-by: Vidya Chidambaram <vidya.chidambaram@dell.com>
Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

mssonicbld commented:
Retrying failed(or canceled) jobs...

vidyac86 committed:
Address FRR unified mode review comments and fix mode consistency issues

Review comment fixes:
- Fix dead-code in docker_init.sh: remove redundant 'separated' condition
  from last elif branch as it's already handled by previous branch
- Add syslog warning for separated mode deprecation to improve visibility
  in 'show logging' output
- Rename frr_mgmt_framework_cfg_template_restore to
  use_template_render_for_restore for better clarity:
  - When true (default): use template rendering for FRR config generation
  - When false: restore from cached DB tables
- Update YANG model, sample configs, and test files to use new variable name
- Document frr.conf.j2 content equivalence in gen_frr.conf.j2 template

Mode consistency fixes:
- Fix supervisord.conf.j2 to handle undefined docker_routing_config_mode
  as unified mode, ensuring vtysh_b runs to replay generated unified config
- Align frrcfgd.py default mode with docker_init.sh: treat missing
  docker_routing_config_mode as unified instead of separated
- Add sonic-cfggen error handling in docker_init.sh before deleting fallback
  config files to prevent losing valid configs if rendering fails

Signed-off-by: Vidya Chidambaram <vidya.chidambaram@dell.com>
Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

mssonicbld commented:
/azp run Azure.sonic-buildimage

vidyac86 committed:
Removed the suspected changes causing failures in checks

Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

vidyac86 committed:
Added a Readme File describing the changes done for this PR for better
understanding

Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

vidyac86 committed:
Revert "Removed a Readme File describing the changes done for this PR for better"

This reverts commit 870f136245aa8361f0f3fa80ebd954e9eabca5a7.

Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

vidyac86 committed:
Fix for the CI failure

Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

vidyac86 committed:
Probable fix for build failure

Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

rookie-who commented:
Hi @vidyac86 — this has approval from @prvattem and is very active. Is this ready to merge? 202605 branch cut is at end of May.

CC: @vaibhavhd @radha-danda (branch managers)

mssonicbld commented:
/azp run Azure.sonic-buildimage

vidyac86 committed:
Update supervisord.conf.j2

Signed-off-by: Vidya Chidambaram <vidya_chidambaram@dell.com>

securely1g reviewed (commented):
Review on behalf of @lguohan:

This PR changes the default `docker_routing_config_mode` from `separated` to `unified` for FRR. A few concerns:

1. **Breaking change** — This changes default behavior for all existing deployments that haven't explicitly set `docker_routing_config_mode`. Have you verified this doesn't break config restore for existing separated-mode deployments during upgrade?
2. **Scope** — The PR has been open for over a year. Is the full changeset still needed, or can it be broken into smaller pieces?
3. **Testing** — What topology/test coverage validates that unified mode works correctly with all FRR daemons (zebra, bgpd, staticd, etc.)?

Please clarify the upgrade path for existing deployments.

mssonicbld commented:
/azp run Azure.sonic-buildimage
